File: server/apps/api-service/src/main/java/com/sharetimer/apiservice/domain/model/Timer.java
package com.sharetimer.apiservice.domain.model;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import lombok.Builder;
import lombok.Getter;

/**
 * Timer domain model
 */
@Getter
@Builder
public class Timer {

  private UUID id;
  private Instant targetTime;
  private UUID ownerToken;
  @Builder.Default
  private List<Timestamp> timestamps = new ArrayList<>();
  private Instant createdAt;
  private Instant updatedAt;

  public void updateTargetTime(Instant targetTime) {
    this.targetTime = targetTime;
  }

  public void updateUpdatedAt(Instant updatedAt) {
    this.updatedAt = updatedAt;
  }

}

File: server/apps/api-service/src/main/kotlin/com/sharetimer/apiservice/domain/model/Timer.kt
package com.sharetimer.apiservice.domain.model

import java.time.Instant
import java.util.UUID

/**
 * Timer domain model
 */
class Timer(
    val id: UUID? = null,
    var targetTime: Instant,
    val ownerToken: UUID? = null,
    val timestamps: MutableList<Timestamp> = mutableListOf(),
    val createdAt: Instant? = null,
    var updatedAt: Instant? = null,
) {
    fun updateTargetTime(
        newTargetTime: Instant,
        requestTime: Instant,
    ) {
        require(requestTime.isAfter(updatedAt ?: Instant.MIN)) { "Stale update request" }
        targetTime = newTargetTime
        updatedAt = requestTime
    }

    fun addTimestamp(capturedAt: Instant) {
        require(!capturedAt.isAfter(targetTime)) {
            "Timestamp must be before target time. capturedAt=$capturedAt"
        }
        timestamps.add(Timestamp(targetTime = targetTime, capturedAt = capturedAt))
    }

    override fun toString(): String =
        "Timer(id=$id, targetTime=$targetTime, ownerToken=$ownerToken, timestamps=${timestamps.size})"
}
